File: src/main/java/de/schramm/royalbash/controller/service/core/creature/CreatureUtil.java
package de.schramm.royalbash.controller.service.core.creature;

import de.schramm.royalbash.controller.service.core.CardOnPlayerContext;
import de.schramm.royalbash.controller.service.core.Creature;
import de.schramm.royalbash.controller.service.core.Game;
import lombok.val;

class CreatureUtil {

    static Game spawnCreature(Creature creature, CardOnPlayerContext cardOnPlayerContext) {

        val game = cardOnPlayerContext.getGame();
        val owner = cardOnPlayerContext.getOwner();
        val targetSpot = cardOnPlayerContext.getTargetSpot();

        val updatedSpot = targetSpot.place(creature);

        return game.findPlayer(owner)
                .map(player -> player.updateSpot(targetSpot, updatedSpot))
                .map(player -> game.updatePlayer(owner, player))
                .orElse(game);
    }

    static boolean isDead(Creature creature) {
        return creature.getHitpoints() <= 0;
    }
}

File: src/main/kotlin/de/schramm/royalbash/domain/card/creature/CreatureUtil.kt
package de.schramm.royalbash.domain.card.creature

import de.schramm.royalbash.domain.Context
import de.schramm.royalbash.domain.Creature
import de.schramm.royalbash.domain.Game

internal object CreatureUtil {

    fun spawnCreature(creature: Creature, context: Context): Game {

        val game = context.game
        val owner = context.owner

        return context.getTargetSpot()
                .map { targetSpot -> game.findPlayer(owner)
                        ?.let { it.updateSpot(targetSpot, targetSpot.place(creature)) }
                        ?.let { game.updatePlayer(owner, it) }
                        ?: game}
                .orElse(game)
    }
}
